### PR TITLE
Android PR 7: Introduce structure of FlutterActivity and FlutterFragment

### PR DESCRIPTION
**Pull request by matthew-carroll**
This PR introduces FlutterActivity and FlutterFragment.  Only the public initialization API is introduced in this PR to keep it brief.  I will add individual responsibilities to each of these classes moving forward.

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  private void ensureFlutterFragmentCreated() {
+    FragmentManager fragmentManager = getSupportFragmentManager();
+    flutterFragment = (FlutterFragment) fragmentManager.findFragmentByTag(TAG_FLUTTER_FRAGMENT);
+    if (flutterFragment == null) {
+      // No FlutterFragment exists yet. This must be the initial Activity creation. We will create
+      // and add a new FlutterFragment to this Activity.
+      flutterFragment = createFlutterFragment();
+      fragmentManager
+          .beginTransaction()
+          .add(FRAGMENT_CONTAINER_ID, flutterFragment, TAG_FLUTTER_FRAGMENT)
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: Can you describe what you mean by multiple instances of `FlutterActivity`s?  Each Activity owns its view hierarchy, so at a minimum it should be true that one of these IDs per Activity is fine.  The only issue would be multiple uses of this ID within the same Activity.

File: shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    }
+  }
+
+  /**
+   * Creates the instance of the {@link FlutterFragment} that this {@code FlutterActivity} displays.
+   * <p>
+   * Subclasses may override this method to return a specialization of {@link FlutterFragment}.
+   */
+  @NonNull
+  protected FlutterFragment createFlutterFragment() {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: In general the point is reasonable.  Can you apply it specifically to Android?  Considering the long history of subclassing Activitys, Fragments, and Views as the primary industry solution for a number of use-cases, can you describe how we can expect to void such situations in general?  This question is especially true of the method that you commented on.  Activitys and Fragments have lifecycles, which means Objects within them must be created and destroyed within that lifecycle.  Do you know how one can achieve that effect without subclassing?  And is that alternative actually simpler and more useful than subclassing?

[the remark this answers is not in the document]
Comment (mklim) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: I wasn't thinking of people extending this and then writing their own app code within it like a typical Android activity/fragment is used, to be honest. I was only thinking of the case of people instantiating this and then bringing it into their app to handle rendering a view within their larger UI. So my thinking was activity is meant to be subclassed, but that doesn't mean that this particular subclass of activity this should also encourage subclassing, especially for methods that already just return a value when it's set in them previously. But if we are expecting/encouraging people to extend this and pile extra logic on top of it in their app inheritance is pretty much unavoidable here.

File: shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mklim) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: I was thinking of multiple activities each with their own view hierarchy--just an abundance of caution. SGTM.

File: shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: You mentioned 2 cases:
 - extending to customize existing behavior
 - extending to introduce new behavior

You said that you were thinking more about number 1.  And you're right, number 1 is what developers are likely to do.  But the question remains, how might a developer effectively achieve number 1 without subclassing?

In the case of providing a custom FlutterEngine, if subclassing is not the primary route then it means developers will have to subclass FlutterActivity to provide a custom engine to the FlutterFragment within the FlutterActivity.  But then, even in that case, the developer is still subclassing FlutterActivity.  So I'm not sure how one achieves composition when using Android framework classes - the mistake appears to exist in the foundation of the framework.

Any thoughts?

[the remark this answers is not in the document]
Comment (mklim) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: So for your specific question it depends on what `FlutterEngine` is being customized to do. Let's say for the sake of argument, somebody wants to always call `bar` when `FlutterEngine` is constructed and `baz` whenever it's destroyed.

```java
// Inheritance example
class MyFlutterEngine extends FlutterEngine {
   MyFlutterEngine(@NonNull Context context, String barParam) {
     super(context);
     bar(barParam);
   }

   @override
   public void destroy() {
     super();
     baz();
   }

   private void bar(String barParam) { /* etc */ }

   private void baz() {/* etc */ }
}
```

So this works fine and looks relatively simple, but unfortunately all the callers of `MyFlutterEngine` need to understand all the concerns of using `FlutterEngine` on top of whatever custom logic that `MyFlutterEngine` adds to it, where `MyFlutterEngine` adds to it. This isn't necessarily going to be surfaced in the subclasses' Java docs, but the lifecycle hooks still are going to need to be wired in for `MyFlutterEngine` the same as its superclass. This is a simplified example so it's fairly manageable, but the complexity gets worse when the chain gets several classes deep and there's multiple state-dependent methods and callbacks to worry about all operating at different levels of the class hierarchy. 

If you wanted to implement that behavior with composition, the trade would be to wrap your instance of `FlutterEngine` in an outer class that blatantly handles and exposes both the custom logic and the logic callers need to manage for the `FlutterEngine` instance within.

```java
// Composition example
final class FlutterEngineBarrer {
   private final FlutterEngine flutterEngine;
   public FlutterEngineBarrer(Context context, String barParam) {
     // Most codebases would use some DI for this instead, but let's pretend we don't have it here.
    flutterEngine = new FlutterEngine(context);
    bar(barParam);
   }

   /** Tear down flutter engine and do (baz logic). */
   public void destroy() {
     flutterEngine.destroy();
     baz();
   }

   private void bar(String barParam) { /* etc */ }

   private void baz() {/* etc */ }

   // Series of wrappers to FlutterEngine

   public void detatchFromJni() {
     flutterEngine.detatchFromJni();
   }

   // etc
}
```

The catch with the above is you still need to pipe any methods from the inner class through, like `detachFromJni()`. But while the boilerplate is annoying in its own right, it does keep the actual contract required for using the class explicit in the class itself and prevents there from being any hidden logic or required info hidden up layers of the class hierarchy. Or require logical deduction on how to use the class based on the interplay of different subclasses in the same class hierarchy. Etc.

Android does heavily encourage inheritance. That doesn't mean that all Android app code needs to also use it for every single utility class by default, but does mean it's unavoidable for some things (such as creating the "main" activity in an app, which does probably apply here).

And again, ultimately this is more of a subjective style thing.

[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/android/FlutterActivity.java]: We spoke offline.  We're in agreement that Java classes should prefer composition, but that Android classes make it impossible.

**Pull request merged**